### PR TITLE
KAFKA-6190 GlobalKTable never finishes restoring when consuming transactional messages

### PR DESCRIPTION
**Pull request by alexjg**
Calculate offset using consumer.position() in GlobalStateManagerImp#restoreState

### CODE REVIEW

**Comment by asfgit** [comment]
FAILURE
 No test results found.
--none--

**Comment by mjsax** [comment]
Thanks for the patch. Can we add a test for this?

**Comment by mjsax** [comment]
\cc @bbejeck @dguy @guozhangwang @rajinisivaram

**Comment by alexjg** [comment]
I had a brief attempt to write a test in `GlobalKTableIntegrationTests` and it looked like I would have to change a fair few other tests to allow creating messages in a transaction with `IntegrationTestUtls`. Does that seem like a reasonable approach? If so I can go ahead and do it.

**Comment by asfgit** [comment]
SUCCESS 
 7148 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 7148 tests run, 5 skipped, 0 failed.
--none--

**Comment by guozhangwang** [comment]
I looked into the consumer internals and verified that we skipped the control messages when setting the position of the consumer. So this fix LGTM.

About the testing coverage: maybe we can do a slight smaller change, on `MockConsumer#addRecords()` to add the batch with controller messages and then in this `GlobalStateManagerImplTest#initializeConsumer` function we can use the modified `addRecords` to add the controller message.

**Comment by mjsax** [comment]
I think it would be good to have an extension of `IntegrationTestUtils` with a boolean flag "enableTransactions" or similar. Internally, we can put the required setting to the producer and add the calls to `initTransactions`, `beginTransaction`, `commitTransaction`. Should not be too hard.

What Guozhang suggested should also work. It's up to you @alexjg

**Comment by alexjg** [comment]
The `enableTransactions` approach was the one I had been thinking of, I'll have a shot at doing that this morning

**Comment by asfgit** [comment]
FAILURE
 No test results found.
--none--

**Comment by alexjg** [comment]
I've added the test as discussed. I ran into a slightly unsatisfying problem with it. The failure mode - which you can trigger by switching the `GlobalStateManagerImpl#restoreState` method back to the previous implementation - seems to cause the entire test harness to just hang rather than timeout with a failure message and continuing with the test suite. I spent some time trying to figure out what was going on but wasn't able to make any progress, any hints on what might be the problem?

**Comment by asfgit** [comment]
SUCCESS 
 7149 tests run, 5 skipped, 0 failed.
--none--

**Comment by dguy** [comment]
Should this be raised against trunk? If the code hasn't changed then we can cherry-pick to 0.11.0, too. Otherwise we'll need another PR for trunk. BTW, thanks @alexjg for picking this up.

**Comment by asfgit** [comment]
SUCCESS 
 7149 tests run, 5 skipped, 0 failed.
--none--

**Comment by dguy** [comment]
@alexjg - the hanging of that method was the original problem that this PR is trying to fix, right? If the test is passing then it would seem you have fixed the problem. To make it fail under circumstances where it hangs you could add a timeout to the `@Test` annotation.

**Comment by alexjg** [comment]
Regarding the test hanging: Firstly I was expecting the code to hang inside the `waitForCondition` call and thus to trigger a failure so I was surprised that this didn't happen. Secondly, I just tried adding a timeout annotation to the testcase and the test case hung indefinitely periodically using 100% CPU. Not sure what is going on there.

Regarding raising against trunk. The `GlobalStateManagerImpl#restoreState` method on `trunk` is slightly different so it would probably make sense to have the fix in a different PR. The tests would be the same though I think.

File: streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java [hunk 1]
```diff
@@ -217,6 +226,37 @@ public boolean conditionMet() {
         }, 30000L, "waiting for final values");
     }
 
+    @Test
+    public void shouldRestoreTransactionalMessages() throws Exception {
+        produceInitialGlobalTableValues(true);
+        startStreams();
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/integration/GlobalKTableIntegrationTest.java:232]: @alexjg The code is actually looping in this line, before the `waitForCondition` call is actually triggered, I think that is why you are seeing the indefinite hanging.

**Comment by guozhangwang** [comment]
LGTM. Merged to 0.11.0, and cherry-picked to 1.0 and trunk.

**Comment by tedyu** [comment]
The following error in trunk seems to be introduced by this PR:
```
<https://builds.apache.org/job/kafka-trunk-jdk8/ws/streams/src/main/java/org/apache/kafka/streams/processor/internals/GlobalStateManagerImpl.java>:195: error: cannot find symbol
                    offset = consumer.position(topicPartition);
                             ^
  symbol:   variable consumer
  location: class GlobalStateManagerImpl
```
I think globalConsumer should be used.

**Comment by guozhangwang** [comment]
Thanks @tedyu ! I have just pushed a hotfix for this.

**Pull request closed without merging**